src/backend/distributed/utils/statistics_collection.c

```diff
@@ -0,0 +1,215 @@
+ *
+ * statistics_collection.c
+ *	  Anonymous reports and statistics collection.
+ *
+ * Copyright (c) 2017, Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#if HAVE_LIBCURL
+
+#include "postgres.h"
+
+#include <curl/curl.h>
+#include <sys/utsname.h>
+
+#include "access/xact.h"
+#include "citus_version.h"
+#include "distributed/metadata_cache.h"
+#include "distributed/statistics_collection.h"
+#include "distributed/worker_manager.h"
+#include "lib/stringinfo.h"
+#include "utils/json.h"
+
+bool EnableStatisticsCollection = true; /* send basic usage statistics to Citus */
+
+static uint64 NextPow2(uint64 n);
+static uint64 ClusterSize(List *distributedTableList);
+static bool SendHttpPostJsonRequest(const char *url, const char *postFields, long
+									timeoutSeconds);
+
+/* WarnIfSyncDNS warns if libcurl is compiled with synchronous DNS. */
+void
+WarnIfSyncDNS(void)
+{
+	curl_version_info_data *versionInfo = curl_version_info(CURLVERSION_NOW);
+	if (!(versionInfo->features & CURL_VERSION_ASYNCHDNS))
+	{
+		ereport(WARNING, (errmsg("your current libcurl version doesn't support "
+								 "asynchronous DNS, which might cause unexpected "
+								 "delays in the operation of Citus"),
+						  errhint("Install a libcurl version with asynchronous DNS "
+								  "support.")));
+	}
+}
+
+
+/*
+ * CollectBasicUsageStatistics sends basic usage statistics to Citus servers.
+ * This includes Citus version, table count rounded to next power of 2, cluster
+ * size rounded to next power of 2, worker node count, and uname data. Returns
+ * true if we actually have sent statistics to the server.
+ */
+bool
+CollectBasicUsageStatistics(void)
+{
+	List *distributedTables = NIL;
+	uint64 roundedDistTableCount = 0;
+	uint64 roundedClusterSize = 0;
+	uint32 workerNodeCount = 0;
+	StringInfo fields = makeStringInfo();
+	struct utsname unameData;
+	memset(&unameData, 0, sizeof(unameData));
+
+	StartTransactionCommand();
+	distributedTables = DistributedTableList();
+	roundedDistTableCount = NextPow2(list_length(distributedTables));
+	roundedClusterSize = NextPow2(ClusterSize(distributedTables));
+	workerNodeCount = ActivePrimaryNodeCount();
+	CommitTransactionCommand();
+
+	uname(&unameData);
+
+	appendStringInfoString(fields, "{\"citus_version\": ");
+	escape_json(fields, CITUS_VERSION);
+	appendStringInfo(fields, ",\"table_count\": " UINT64_FORMAT, roundedDistTableCount);
+	appendStringInfo(fields, ",\"cluster_size\": " UINT64_FORMAT, roundedClusterSize);
+	appendStringInfo(fields, ",\"worker_node_count\": %u", workerNodeCount);
+	appendStringInfoString(fields, ",\"os_name\": ");
+	escape_json(fields, unameData.sysname);
+	appendStringInfoString(fields, ",\"os_release\": ");
+	escape_json(fields, unameData.release);
+	appendStringInfoString(fields, ",\"hwid\": ");
+	escape_json(fields, unameData.machine);
+	appendStringInfoString(fields, ",\"api_version\": ");
```

Comment:
Oh, I somehow missed this in a previous review.

The version is not part of the request. It's metadata.

There are [various arguments](https://www.troyhunt.com/your-api-versioning-is-wrong-which-is/) about different ways of versioning an API, but I don't think "include it in the request" is usually considered at all.

If you're using e.g. Django, [this library](http://www.django-rest-framework.org/api-guide/versioning/) appears to support many of the common methods. I looped in some of the cloud folk for their opinion on this topic…

Basically a lot of the chatter online is "headers are most semantic because they don't pollute the URL". But they're also touchier, harder to test with e.g curl, more verbose, have issues with badly configured caching proxies/CDNs, etc. I think we should just put a version in our URL as a top-level path component.

Which brings up my final point: the `STATS_COLLECTION_URL` will probably host a version-check API at some point as well. It would be best to `#define` the host name and add a parameter to `SendHttpPostJsonRequest` to allow passing an arbitrary path.

To sum up, in this case I think we should be `POST`ing our JSON to `/v1/usage_reports` (since the thing you're creating is a usage report, so that naming is in line with RESTy practices for creating a new object over HTTP).